### ISSUE TITLE
`Can't change a seen value` error when running certain query

### ISSUE THREAD

**Issue by alexche-cchange** [neo4j/neo4j#13348]
In some situations, I am getting a cryptic `Can't change a seen value` error from Neo4j. The minimal sample query is below. All the parts of the query are important, so reducing it further makes issue disappear. The error manifests before the query execution: `EXPLAIN`ing the same query also results in an error. 

```
MATCH (node:A)
WHERE (node:B OR node:C)

WITH DISTINCT node

RETURN node

ORDER BY node.prop ASC
```

Unfortunately, I am stably reproducing this on our production graph which I can not share, but I **could not** create a set of sample data to trigger the same. Our prod DB is fairly complex with ~700 unique labels, ~200K of nodes of type A, and deeply connected graph. 

Removing any part of the query makes error disappear: 
- `OR` between the secondary labels
- `WITH DISTINCT`, and also moving `DISTINCT` to `RETURN` clause
- Removing sorting altogether, or sorting by something else joined to `A`

Also, in a real query, we do also return joined nodes. If I add an explicit hint `USING JOIN ON other_node`, error disappears. 

If I remove the sorting clause, for example, and run `EXPLAIN`, this is a plan I get - it starts with two scans followed by sorted merge.

![plan_minimal](https://github.com/neo4j/neo4j/assets/120404756/13e442f1-34db-4d64-83a4-4f633b1d38a4)

Maybe the fact that the result set has been already sorted somehow triggers planner when it sees the resulting `ORDER BY`? 

**Neo4j Version:** 4.4.27
**Operating System:** Arch/6.6 kernel 
**API:** Docker neo4j:4.4.27-enterprise. Issue manifests in built-in web console at localhost:7474

### Steps to reproduce
1. Run the query on our big prod graph
2. Run the query via built-in web console at localhost:7474

### Expected behaviour
Query would be planned and executed.

### Actual behaviour
Query is not planned neither executed.

**Comment by gem-neo4j** [comment]
Hi!

Thanks for reporting, we'll take a look :)

**Comment by hedengran** [comment]
Hi @alexche-cchange ! Thank you again for reporting this. I understand that you can't share your production database, but would it be possible for you to export graph counts for us? 

```
CALL db.stats.retrieve("GRAPH COUNTS")
```

It would help us a lot with reproducing the error.

**Comment by hedengran** [comment]
Alternatively, if it's more appropriate to provide anonymized graph counts you could use:

```
db.stats.retrieveAllAnonymized("GRAPH COUNTS")
```

It works a bit different, so you will get a lot more output. But everything you need to copy and provide here is the row where `section = "GRAPH COUNTS"`.

**Comment by alexche-cchange** [comment]
Thank you @hedengran!

Attaching the graph counts (anon) here. 

[for_neo4j_graph_counts.txt](https://github.com/neo4j/neo4j/files/13429041/for_neo4j_graph_counts.txt)

**Comment by hedengran** [comment]
Hello @alexche-cchange ! Thank you for the graph counts, and for the detailed original report. I was able to reproduce the issue and it was indeed a very specific combination of plans that triggered the case. I've merged a fix in 4.4, and it will be present in the next release (4.4.29). I've also merged the same fix in 5.x, however, worth noting is that this specific issue would not manifest in 5.x because of some improvements elsewhere. 

Did I understand correctly that the hint `USING JOIN ON other_node` is a sufficient workaround for now? If you have any other questions let us know!

**Comment by alexche-cchange** [comment]
Hi @hedengran, thank you very much! 

> Did I understand correctly that the hint USING JOIN ON other_node is a sufficient workaround for now?

Actually, the change in query which lead to an issue was an incremental update we are able to hold for a while. `USING JOIN ON other_node` did make the problem go, but caused quite a hefty performance penalty, so we did not go with it to production.

So, all-in-all, we would be happy to wait till 4.4.29 maintenance is out and will test it then. 

Again, kudos for picking this up. 

Regards,
Alex.
